### ISSUE TITLE
core.bitop.bsf not callable using ulong

### ISSUE THREAD

**Issue by JudgeFish** [ldc-developers/ldc#2926]
I encountered the following error when compiling ldc "master" using a freshly built ldc "ltsmaster":

ldc/dmd/builtin.d(254): Error function core.bitop.bsf (uint v) is not callable using argument types (ulong)
(and corresponding for bcf)

System:
Rasbery Pi 3B+ Arch Linux 4.14.31
LDC 0.17.6
LLVM 7.0.0
armv71-unknown-linux-gnueabihf
cortex-a53

Note: Patching the sources locally with an ugly cast it was possible to compile the rest.

**Comment by kinke** [comment]
Ah, seems like master (probably since v2.083 or 2.082) isn't compilable with a 32-bit ltsmaster host compiler (not just ARM), because ltsmaster's druntime only provides `bs{f,r}(size_t)`.

**Comment by kinke** [comment]
Yeah I've always been reluctant wrt. backporting stuff to ltsmaster. As it only concerns 32-bit host compilers, I don't really care. ;)

> would be nice if we could move all CTFE reals to IEEE Quadruple precision

Sounds like a reasonable choice, precision-wise. AFAIK, GDC uses a software `real_t` type on all hosts, so I guess we don't have to care much about compile-time performance.

> Only issue recently has been lower-precision CTFE reals on host screwing up some cross-compiled stdlib tests on target

Hmm, could you try rebuilding LDC yet another time, with the LDC generated by the cross-compiler? I'd have hoped that overflows and inaccuracies don't propagate from the host to the target, at least not across 2 hops. Your current final LDC probably still features wrong values for embedded constants, as those have been computed by the cross-compiler with limited 80-bit `real_t` precision.

**Comment by joakim-noah** [comment]
>I'd have hoped that overflows and inaccuracies don't propagate from the host to the target, at least not across 2 hops

No, you're reading too far into it, that's fine, as the DMD frontend doesn't really use druntime/Phobos much and I doubt it has any constant reals.

I was talking strictly about the first porting step, ie cross-compiling the test runners from x64 to a platform with IEEE Quadruple real precision.

**Comment by kinke** [comment]
> No, you're reading too far into it, that's fine, as the DMD frontend doesn't really use druntime/Phobos much and I doubt it has any constant reals.

Ah okay, all good then. The CTFE stuff switched back to the C runtime; we've been using host Phobos for some LDC versions. While I don't like that C/Phobos divergence (slightly different CTFE/runtime results quite possible), it makes bootstrapping a lot easier.
There are some `real` constants, e.g., FP type properties (`T.{nan,max,min,epsilon,...}`), but at least those are populated via LLVM's `APFloat` (and likely over/underflow when cross-compiling to a target with higher precision).

**Comment by JohanEngelen** [comment]
@JudgeFish About the datetime issue: the file std/datetime.d no longer exists. So it should not be there in your source tree. Probably something went wrong with your git checkout such that old files are lingering there.

**Comment by kinke** [comment]
That's most likely caused by installing different versions of LDC into the same directory, without removing it first - the imports are simply copied there, overwriting previous versions, but if something (`std.datetime`) turns from monster module to package, you get these problems; I had them too. ;)

**Comment by JudgeFish** [comment]
Yes, that was the cause... I just used "make install" back-to-back.
Thanks

**Comment by Apteryks** [comment]
Still the case on 1.27.1 (tested: i686-linux and armhf-linux).
